snuba/state/cache/redis/backend.py

```diff
@@ -0,0 +1,209 @@
+        # the client responsible for executing the function and notifying the
+        # subscribed clients of its completion. Only one wait queue should be
+        # associated with a cache key at any time.
+        wait_queue_key = self.__build_key(key, "tasks", "wait")
+
+        # The task identity (a Redis bytestring) is used to store a unique
+        # identifier for a single task evaluation and notification cycle. Only
+        # one task should be associated with a cache key at any time.
+        task_ident_key = self.__build_key(key, "tasks")
+
+        # The notify queue (a Redis list) is used to unblock clients that are
+        # waiting for the task to complete. **This implementation requires that
+        # the number of clients waiting for responses via the notify queue is
+        # no greater than the number of clients in the wait queue** (minus one
+        # client, who is doing the work and not waiting.) If there are more
+        # clients waiting for notifications than the members of the wait queue
+        # (for any reason), some set of clients will never be notified. To be
+        # safe and ensure that each client only waits for notifications from
+        # tasks where it was also a member of the wait queue, the notify queue
+        # includes the unique task identity as part of it's key.
+        def build_notify_queue_key(task_ident: str) -> str:
+            return self.__build_key("key", "tasks", f"notify/{task_ident}")
+
+        # At this point, we have all of the information we need to figure out
+        # if the key exists, and if it doesn't, if we should start working or
+        # wait for a different client to finish the work. We have to pass the
+        # task creation parameters -- the timeout (execution deadline) and a
+        # new task identity just in case we are the first in line.
+        result = self.__script_get(
+            [result_key, wait_queue_key, task_ident_key], [timeout, uuid.uuid1().hex]
+        )
+
+        if result[0] == RESULT_VALUE:
+            # If we got a cache hit, this is easy -- we just return it.
+            logger.debug("Immediately returning result from cache hit.")
+            return self.__codec.decode(result[1])
+        elif result[0] == RESULT_EXECUTE:
+            # If we were the first in line, we need to execute the function.
+            # We'll also get back the task identity to use for sending
+            # notifications and approximately how long we have to run the
+            # function. (In practice, these should be the same values as what
+            # we provided earlier.)
+            task_ident = result[1].decode("utf-8")
+            task_timeout = int(result[2])
+            logger.debug(
+                "Executing task (%r) with %s second timeout...",
+                task_ident,
+                task_timeout,
+            )
+
+            argv = [task_ident, 60]
+            try:
+                value = function()
```

Comment:
Should this also be subject to timeout? (or maybe timeout called something else) It seems kind of unexpected for a timeout argument to only be applied for subsequent tasks in the queue but ignored for the first one.

Reply:
Hmm, this is a good point and something that I hadn't really considered.

We don't really have a way to reliably time out an arbitrary function without pushing it into a different thread — that might make sense here, especially since the client who sent the request to us may have timed out the request on their side anyway.

My only concern (if we were to use a thread pool here) would be that we can end up in a situation where all of the available threads are occupied doing work that we have already passed on the result (we have no way to kill threads once they are no longer needed — really bad if one enters an infinite loop or something) and we no longer have the ability to run new tasks… but I guess that's not really significantly different than running out of threads/processes to handle web requests that have all been already assumed to have timed out by the client on the other end of the connection?

I will give this some more thought.

Reply:
Here is the hypothetical change: a23a52c1e81b052df7f1d5dac2fc1b6743b228d3

I don't have a strong opinion one way or other the about this, and feel like I could honestly make reasonable arguments in good faith both for and against merging it into this change. Thoughts?

Reply:
Merged that in after discussing out of band. Thanks for pointing out the inconsistency.